src/input/scroll.c

```diff
@@ -242,131 +192,87 @@ void scroll_set_custom_margins(int x, int y, int width, int height)
 
 void scroll_restore_margins(void)
 {
-    data.is_touch = 0;
     data.limits.active = 0;
 }
 
-touch_coords scroll_get_original_touch_position(void)
+void scroll_drag_start(int is_touch)
 {
-    return data.start_touch;
+    if (data.drag.active) {
+        return;
+    }
+    data.drag.active = 1;
+    data.drag.is_touch = is_touch;
+    data.drag.delta.x = 0;
+    data.drag.delta.y = 0;
+    if (!is_touch) {
+        system_mouse_get_relative_state(0, 0);
+    }
+    clear_scroll_speed();
 }
 
-void scroll_start_mouse_drag(const mouse *m, const pixel_offset *position)
+void scroll_drag_move(void)
 {
-    data.drag.has_scrolled = 0;
-    data.drag.start_x = m->x;
-    data.drag.start_y = m->y;
-    data.is_scrolling = 1;
-    data.is_touch = 0;
-    clear_scroll_decay(position);
-}
+    if (!data.drag.active) {
+        return;
+    }
 
-int scroll_move_mouse_drag(const mouse *m, pixel_offset *position)
-{
-    if (!data.is_scrolling) {
-        return 0;
+    int delta_x = 0;
+    int delta_y = 0;
+    if (!data.drag.is_touch) {
+        system_mouse_get_relative_state(&delta_x, &delta_y);
+    } else {
+        const touch *t = get_earliest_touch();
+        delta_x = -t->frame_movement.x;
+        delta_y = -t->frame_movement.y;
     }
 
-    if (!data.drag.has_scrolled) {
-        int delta_x = abs(m->x - data.drag.start_x);
-        int delta_y = abs(m->y - data.drag.start_y);
-        if (delta_x <= MOUSE_SCROLL_MIN_DELTA && delta_y <= MOUSE_SCROLL_MIN_DELTA) {
-            return 0;
+    data.drag.delta.x += delta_x;
+    data.drag.delta.y += delta_y;
+    if ((delta_x != 0 || delta_y != 0)) {
+        if (!data.drag.is_touch) {
+            system_mouse_set_relative_mode(1);
+        }
+        // Store tiny movements until we decide that it's enough to move into scroll mode
+        if (!data.is_scrolling) {
+            data.is_scrolling = abs(data.drag.delta.x) > SCROLL_DRAG_MIN_DELTA || abs(data.drag.delta.y) > SCROLL_DRAG_MIN_DELTA;
+        }
+    }
+    if (data.is_scrolling) {
+        time_millis current_time = time_get_millis();
+        time_millis delta_time = current_time - data.last_time;
+
+        // SDL touch events sometimes incorrectly set movement to zero before the finger is released
+        // This gives a time tolerance before setting the movement to 0
+        if (!data.drag.is_touch ||
+            data.drag.delta.x != 0 || data.drag.delta.y != 0 ||
+            delta_time > DRAG_STOPPED_TIME_TOLERANCE) {
+            data.last_time = current_time;
```

Comment:
I had a better look at this, this isn't an SDL bug, but my own, in `platform/touch.c`. So don't merge until I properly fix this, please.